src/bynder-js-sdk.js

```diff
@@ -571,4 +573,327 @@ export default class Bynder {
+     */
+    saveAsset(data) {
+        if (!this.validURL()) {
+            return rejectURL();
+        }
+        const { brandId, mediaId } = data;
+        if (!brandId || brandId === '') {
+            return rejectValidation('upload', 'brandId');
+        }
+        const saveURL = mediaId ? `v4/media/${mediaId}/save/` : 'v4/media/save/';
+        const request = new APICall(
+            this.baseURL,
+            saveURL,
+            'POST',
+            this.consumerToken,
+            this.accessToken,
+            data
+        );
+        return request.send();
+    }
+
+    /**
+     * Uploads an arbitrarily sized buffer or stream file and returns the uploaded asset information
+     * @see {@link https://bynder.docs.apiary.io/#reference/upload-assets}
+     * @param {Object} file={} - An object containing the id of the desired collection.
+     * @param {String} file.filename - The file name of the file to be saved
+     * @param {Buffer|Readable} file.body - The file to be uploaded. Can be either buffer or a read stream.
+     * @param {Number} file.length - The length of the file to be uploaded
+     * @param {Object} file.data={} - An object containing the assets' attributes
+     * @return {Promise} The information of the uploaded file, including IDs and all final file urls.
+     */
+    uploadFile(file) {
+        const { body, filename, data } = file;
+        const { brandId } = data;
+        const isBuffer = Buffer.isBuffer(body);
+        const isStream = typeof body.read === 'function';
+        const length = isBuffer ? body.length : file.length;
+
+        if (!this.validURL()) {
+            return rejectURL();
+        }
+        if (!brandId || brandId === '') {
+            return rejectValidation('upload', 'brandId');
+        }
+        if (!filename || filename === '') {
+            return rejectValidation('upload', 'filename');
+        }
+        if (!body || (!isStream && !isBuffer)) {
+            return rejectValidation('upload', 'body');
+        }
+        if (!length || typeof length !== 'number') {
+            return rejectValidation('upload', 'length');
+        }
+
+        const getClosestUploadEndpoint = this.getClosestUploadEndpoint.bind(this);
+        const initUpload = this.initUpload.bind(this);
+        const registerChunk = this.registerChunk.bind(this);
+        const finalizeUpload = this.finalizeUpload.bind(this);
+        const pollUploadStatus = this.pollUploadStatus.bind(this);
+        const saveAsset = this.saveAsset.bind(this);
+
+        return Promise.all([
+            getClosestUploadEndpoint(),
+            initUpload(filename)
+        ])
+        .then((res) => {
+            const [endpoint, init] = res;
+            const uploadPath = init.multipart_params.key;
+
+            // count number of chunks
+            const CHUNK_SIZE = 1024 * 1024 * 5;
+            const chunks = Math.ceil(length / CHUNK_SIZE);
+
+            const uploadChunkToS3 = (chunkData, chunk) => {
```

Comment:
This function should be extracted to a class method.

Reply:
Not sure how to do this efficiently. Perhaps you have a suggestion?